### PR TITLE
Fixes #25179, refactored Redis Lookup

### LINKED ISSUE #25179: Redis Lookup Plugin refactor
##### ISSUE TYPE

 - Feature Idea

##### COMPONENT NAME
- Redis Lookup Plugin

##### ANSIBLE VERSION

```
ansible 2.3.0.0
  config file = /etc/ansible/ansible.cfg
  configured module search path = Default w/o overrides
  python version = 2.7.13 (default, May 10 2017, 20:04:36) [GCC 6.3.1 20161221 (Red Hat 6.3.1-1)]

```

##### CONFIGURATION
N/A

##### OS / ENVIRONMENT
Linux

##### SUMMARY

Reasons for this enhancement:

- **redis_kv** Lookup Plugin are a bit old and need a refactor
- Receive an array and returns all the values required on this array
- Receive an string and returns the value associated to the Key 
- Environment variables for IP and Port Server

##### STEPS TO REPRODUCE

- Input

```yaml
- name: '[CFG MGMT] Recover the file status from storage'
  set_fact:
    redis_multi_var: "{{ lookup('redis_kv', 'key1', 'key2', 'key3') }}"
    redis_array: "{{ lookup('redis_kv', files) }}"

- debug: var=redis_multi_var
- debug: var=redis_array
```

##### EXPECTED RESULTS
```
PLAY [Ansible Configuration Management] ******************************************************************************************************************************************************************************************************

TASK [Gathering Facts] ***********************************************************************************************************************************************************************************************************************
ok: [localhost]

TASK [ansible-cfg-mgmt : [VALIDATIONS] Checking Ansible variables for this role] *************************************************************************************************************************************************************
skipping: [localhost] => (item=key1) 
skipping: [localhost] => (item=key2) 
skipping: [localhost] => (item=key5) 

TASK [ansible-cfg-mgmt : [CFG MGMT] Recover the file status from storage] ********************************************************************************************************************************************************************
ok: [localhost]

TASK [ansible-cfg-mgmt : debug] **************************************************************************************************************************************************************************************************************
ok: [localhost] => {
    "changed": false, 
    "redis_multi_var": "Value1,Value2,Value3"
}

TASK [ansible-cfg-mgmt : debug] **************************************************************************************************************************************************************************************************************
ok: [localhost] => {
    "changed": false, 
    "redis_array": "Value1,Value2,Value5"
}

PLAY RECAP ***********************************************************************************************************************************************************************************************************************************
localhost                  : ok=4    changed=0    unreachable=0    failed=0   
```
##### ACTUAL RESULTS
N/A

### PR DESCRIPTION
**Pull request by jparrill**
##### SUMMARY
Fixes #25179 

This refactor will take care of those enhancements:

- redis_kv Lookup Plugin are a bit old and need a refactor
- Receive an array and returns all the values required on this array
- Receive an string and returns the value associated to the Key
- Environment variables for IP and Port Server

New Environment Variables with default values:

- **ANSIBLE_REDIS_URL:** 127.0.0.1
- **ANSIBLE_REDIS_PORT:** 6379

##### ISSUE TYPE
 - Feature Pull Request

##### COMPONENT NAME
**redis_kv** Lookup Plugin

##### ANSIBLE VERSION

```
ansible 2.3.0.0
  config file = /etc/ansible/ansible.cfg
  configured module search path = Default w/o overrides
  python version = 2.7.13 (default, May 10 2017, 20:04:36) [GCC 6.3.1 20161221 (Red Hat 6.3.1-1)]

```

##### ADDITIONAL INFORMATION
This change is necessary because you cannot pass an array with a group of keys to recover, before this changes you need to concat the server url and a bunch of keys.

Now the URL and Port are Environment variables and it's more clean to claim a variable from redis without the server in the lookup call.

- Input YML file:

```yml
---
- name: 'Ansible Configuration Management'
  hosts: localhost
  connection: local
  vars:
    entry_files: ['key1','key2','key5']
  tasks:
    - name: '[CFG MGMT] Recover the file status from storage'
      set_fact:
        redis_multi_var: "{{ lookup('redis_kv', 'key1', 'key2', 'key3') }}"
        redis_array: "{{ lookup('redis_kv', entry_files) }}"

    - debug: var=redis_multi_var
    - debug: var=redis_array
```

- Output:

```
PLAY [Ansible Configuration Management] ******************************************************************************************************************************************************************************************************

TASK [Gathering Facts] ***********************************************************************************************************************************************************************************************************************
ok: [localhost]

TASK [[CFG MGMT] Recover the file status from storage] ***************************************************************************************************************************************************************************************
ok: [localhost]

TASK [debug] *********************************************************************************************************************************************************************************************************************************
ok: [localhost] => {
    "changed": false, 
    "redis_multi_var": "Value1,Value2,Value3"
}

TASK [debug] *********************************************************************************************************************************************************************************************************************************
ok: [localhost] => {
    "changed": false, 
    "redis_array": "Value1,Value2,Value5"
}

PLAY RECAP ***********************************************************************************************************************************************************************************************************************************
localhost                  : ok=4    changed=0    unreachable=0    failed=0   
```

### CODE REVIEW

File: lib/ansible/plugins/lookup/redis_kv.py [hunk 1]
```diff
@@ -14,59 +15,116 @@
...
-    import redis        # https://github.com/andymccurdy/redis-py/
-    HAVE_REDIS=True
+    import redis
 except ImportError:
-    pass
+    raise AnsibleError("Can't LOOKUP(redis_kv): module redis is not installed")
 
-from ansible.errors import AnsibleError
-from ansible.plugins.lookup import LookupBase
+if os.getenv('ANSIBLE_REDIS_URL') is not None:
```
Comment (bcoca) [inline lib/ansible/plugins/lookup/redis_kv.py]: wny not just ```myvar = os.getenv('var', 'default')```?

**Review by bcoca — CHANGES_REQUESTED** [review]

File: lib/ansible/plugins/lookup/redis_kv.py (hunk 1 above)
Comment (jparrill) [reply lib/ansible/plugins/lookup/redis_kv.py]: More elegant, implementing it ;)

**Comment by jparrill** [comment]
I think that this error are a platform one :D, on the last CI tests
`ReadTimeout: HTTPSConnectionPool(host='ec2-13-58-110-185.us-east-2.compute.amazonaws.com', port=5986): Read timed out. (read timeout=30)`

I already make a new push to relaunch it

**Comment by jparrill** [comment]
I think there is some problems with the deployment platform, isn't it?

**Comment by gundalow** [comment]
I've triggered a rerun of Shippable.
For future reference you can do this by closing & opening the PR

**Comment by jparrill** [comment]
It's Ok for you @bcoca ?

**Comment by gundalow** [comment]
closing and reopening to trigger CI

**Comment by bcoca** [comment]
i hope #33883 supersedes this one

**Comment by Akasurde** [comment]
Closing this issue as PR #33883 is merged and supersedes this.

**Pull request closed without merging**